### PR TITLE
Add systemd-resolved DNS forwarding guide

### LINKED ISSUE #4155: DNS forwarding with dnsmasq under systemd
#### Overview of the Issue

The consul guides explain [how to run dnsmasq as root, forwarding certain requests to consul](https://www.consul.io/docs/guides/forwarding.html#dnsmasq-setup). This worked well before systemd (e.g. ubuntu 14.04 and 16.04), but with more recent linux distros running systemd (ubuntu 17.04 and 18.04 LTS) the current guide won't work.

systemd-resolved will control `/etc/resolv.conf` (it's just a symlink now), so dig will use systemd-resolved by default, instead of dnsmasq.

I see two solutions here:

1. Find a way to have dnsmasq intercept DNS-queries, pipe the relevant ones to consul and pass along the others to systemd-resolved, without having to disable systemd-resolved.
2. Sidestep dnsmasq and have systemd-resolved somehow pass along relevant dns queries along to consul.
3. Explain how to completely disable systemd-resolved and run something similar to what the current guides suggest (I think this option is the least desirable). 

All three of them would require guide updates. A related issue is this one (https://github.com/hashicorp/consul/issues/3945).

***

One partial workaround is to configure the following:

```
# /etc/systemd/resolved.conf
[Resolve]
DNS=127.0.0.1
# …
```

In that case resolved will ask dnsmasq (indirectly consul) simultaneously with asking its upstream dns (usually from dhcp), which seems to work. However, that would leak consul dns queries outside the local machine. It will probably also be brittle, i.e. it relies on the response from dnsmasq coming in before those from the per-link dns servers.

> DNS requests are sent to one of the listed DNS servers in parallel to suitable per-link DNS servers acquired from systemd-networkd.service(8) or set at runtime by external applications.
> — source: https://www.freedesktop.org/software/systemd/man/resolved.conf.html

This leaking feels like a rather big downside.

***

#### Reproduction Steps

1. Run a vanilla ubuntu 18.04 machine
2. Setup consul
3. Follow the dns [forwarding guide](https://www.consul.io/docs/guides/forwarding.html#dnsmasq-setup)

### PR DESCRIPTION
**Pull request by mkeeler**
Fixes #4155

Added a section to the docs on how to forward DNS queries for the consul domain to Consul using systemd-resolved configuration.

### CODE REVIEW

File: website/source/docs/guides/forwarding.html.md [hunk 1]
```diff
@@ -155,6 +156,32 @@ You may have to add the following line to the bottom of your
 include: "/etc/unbound/unbound.conf.d/*.conf"
 ```
 
+### systemd-resolved Setup
+
+systemd-resolved is typically configured with `/etc/systemd/resolved.conf`. 
+To configure systemd-resolved to send queries for the consul domain to
+Consul, configure resolved.conf to contain the following:
```
Comment (banks) [inline website/source/docs/guides/forwarding.html.md]: nitpick: I'd put `systemd-resolved` in code fences as it reads a bit funny if it's not clearly the name of a binary. Might not be consistent with other docs here I guess...

File: website/source/docs/guides/forwarding.html.md [hunk 2]
```diff
@@ -155,6 +156,32 @@ You may have to add the following line to the bottom of your
...
+cannot contain ports. So for this to work either Consul must be
+[configured to listen on port 53](https://www.consul.io/docs/agent/options.html#dns_port)
+instead of 8600 or you can use iptables to map port 53 to 8600. 
+The following iptables commands are sufficient to do the port
+mapping.
+
+```
+[root@localhost ~]# iptables -t nat -A OUTPUT -d localhost -p udp -m udp --dport 53 -j REDIRECT --to-ports 8600
+[root@localhost ~]# iptables -t nat -A OUTPUT -d localhost -p tcp -m tcp --dport 53 -j REDIRECT --to-ports 8600
+```
```
Comment (sandstrom) [inline website/source/docs/guides/forwarding.html.md]: How about using the expanded flags, making it easier to understand.

```
iptables --table nat --append OUTPUT --protocol udp --destination localhost --match udp --dport 53 --jump REDIRECT --to-ports 8600
```

1. Also, I'm curious if the `--match` flag is needed?
2. Would there be any benefit to adding `--out-interface lo`, to restrict this to the loopback interface?

Comment (mkeeler) [reply website/source/docs/guides/forwarding.html.md]: So all I did was copy two of the necessary iptables rules from the iptables section of the forwarding guide which use the shortened -t -A -d etc.

As for the --match and --out-interface: --out-interface should be redundant as the destination being matches is localhost. Unless you have configured localhost to map to another IP or are doing something very wrong with your routing this should be adequate to restrict it to the loopback interface. The -m flags are definitely needed in order to only forward 53 to 8600 and not all UDP traffic.

Comment (sandstrom) [reply website/source/docs/guides/forwarding.html.md]: Sounds good!

File: website/source/docs/guides/forwarding.html.md [hunk 3]
```diff
@@ -155,6 +156,32 @@ You may have to add the following line to the bottom of your
 include: "/etc/unbound/unbound.conf.d/*.conf"
 ```
 
+### systemd-resolved Setup
```
Comment (timglabisch) [inline website/source/docs/guides/forwarding.html.md]: doesn't fully work for me, i also have to

```
sudo rm -rf /etc/resolv.conf

change  /etc/resolv.conf to
nameserver 127.0.0.1
nameserver 127.0.0.53
```

**Comment by otrapeznikov** [comment]
Do you guys have any documentation on how to make resolve reverse lookup queries by external DNS only? After following the guid from this PR, my node is resolving it's FQDN by two different name servers, depending which one was picked by systemd-resolved.

**Comment by banks** [comment]
Hey @otrapeznikov asking questions on old closed PRs can mean they get lost. I suggest you head over to https://discuss.hashicorp.com/c/consul to ask there and get attention from the Consul community and engineers! Thanks!

**Pull request merged**